### DISCUSSION TITLE
Syncing by resampling / sample insertion

### DISCUSSION THREAD

**Discussion by CarlosDerSeher** [CarlosDerSeher/snapclient#58] [Ideas]
I talked to the developer of arduino snapclient and he keeps the sync by resampling. I like the concept very much but couldn't find an easy way to do it in C on the esp32. We can't use ADF because of closed source flac used by their lib.

Any ideas?

**Comment by DerPicknicker** [comment]
Dumb question: 

Copy the method of the Arduino code. If the license matches? Why not. your code is GPL3 so for business purposes it’s not an option. 

I don’t see any problems.

Could you be a little more precise what do you exactly mean with re-sample? Does the DSP-Library of espressif offer such a functionality?

**Reply by CarlosDerSeher** [reply]
Yes I could probably port the arduino library from C++ to C.
And I disagree, you could still make a product with GPL licensed code.

Esp-dsp doesn't give you a resample filter out of the box but probably all the tools to build one I guess. It's been a while since I've been working with DSPs.

I think it works like this: set i2s playback to use 44100Hz. If we are early slow down playback by resampling a chunk of 20ms to 44200Hz. You get more samples that way but keep playing them with 44100Hz. Same goes for the other direction. Maybe don't resample that extreme but only by 1Hz or so...

**Reply by DerPicknicker** [reply]
Regarding GPL3 it's possible yes - but nobody would do this. You have to publish a lot of your own code which I really don't want. 

But in general what's the background of your question? Is it regarding ESPHOME or just for improving this project? 

I tested the current version and I really like the sync. For me it was not audible even if I sit only 10cm away from the speakers. If it's regarding ESPHOME I can understand why you're interested in a more arduino compatible way. 

In general I would suggest using a ring-buffer which is really efficient. And would allow to easily drop out old data. The resampling should work with it fine if I understood the concept right.

One thing might be that it can cause some audio quality topics. A lot of resampling is not recommended in terms of sound quality.


EDIT: See here: https://github.com/espressif/esp-adf/tree/dd204c4/examples/audio_processing/pipeline_resample 

Espressif already has some resampling examples.. But if you see there are 100+ open issues I would stay away from Adf. The integration into IDF is not that great and the documentation doesn't help sometimes. I can understand why you removed everything except the codecs......

**Reply by CarlosDerSeher** [reply]
> You have to publish a lot of your own code which I really don't want.

The code is already public. But you could build some bad ass good looking high quality audio player hardware and sell that as open source multi room system. Not that many people can design a good quality audio PCB (which you wouldn't need to open source). FreeNas for example, they sell high end server products and their OS is fully open source.

> But in general what's the background of your question? Is it regarding ESPHOME or just for improving this project?

It's mostly because of #55. This got me thinking about alternatives. 

> One thing might be that it can cause some audio quality topics. A lot of resampling is not recommended in terms of sound quality.

Yes that's why haven't looked into it so much yet. Maybe a better approach would be to just stuff in/remove some samples to get sync to the right position without apll tuning. Maybe I'll give this a try again some day.

> Espressif already has some resampling examples..

But they use a library which contains the resample filter and unfortunately also contains flac and I really don't want to use that pipeline crap. I tried that in the very beginning of the project and it was a dead end for me.

**Reply by DerPicknicker** [reply]
> It's mostly because of #55. This got me thinking about alternatives.

I think as said in the issue it's hardware caused. I use here an 21$ board with 2x23W and even connected to a sub with a bad PSU it sound totally normal. Even at 60% which is already really loud. I think it's not a software problem in that case. 
My code is by far not the best - and even there I can't say that the sound quality is a problem. My board has some capacitors to reduce noise of psu. 

Could you briefly describe the difference between resampling and your current implementation?

**Reply by CarlosDerSeher** [reply]
Resampling would result in more or less samples which will be played back at the original sample rate. Resulting in slower or faster playback respectively. MCLK will always stay the same.

My implementation uses apll and I increase / decrease this clock just a tiny bit to keep the sync in a tight threshold.

My goal was to get the sync nearly perfect so even wireless stereo speakers would be possible. I didn't try that but I think I missed that goal. Still for multi room this works very very good.

Another advantage of changing it to resampling or sample stuffing would be compatibility with esp32s without apll.

**Reply by DerPicknicker** [reply]
> My goal was to get the sync nearly perfect so even wireless stereo speakers would be possible.

Don't be too hard - I tested it. For me it works fine. One speaker was an Android client and the other one the esp32 for me it was the perfect wireless stereo setup. 

> Resampling vs. APLL 

From my understanding Resampling requires more cpu-time and more ram. Because you have to save the re-sampled data somehow or you can do the resampling directly on the i2s hardware. Because you just change the sample rate of the i2s component.

**Reply by CarlosDerSeher** [reply]
Glad to read that. Still I have to measure how much it is off. I read it should be in the range of 10us to 100us max for it not to be audible

**Reply by CarlosDerSeher** [reply]
@DerPicknicker 
I've pushed a new branch https://github.com/CarlosDerSeher/snapclient/commit/1e3bda9ad01025e566f8aadd6a705e0a76ba73b2

Where I've implemented a different syncing approach. It's still experimental, maybe you are willing to test

**Reply by DerPicknicker** [reply]
@CarlosDerSeher ... Yeah I will test it. But it will take some time. First I want to finish the TAS5805M driver and then I will try it. 

Does it use IDF5 or 4..?

**Reply by CarlosDerSeher** [reply]
V5

**Reply by DerPicknicker** [reply]
I will try it next week. I am really busy these days. 

What's your impression about the re-sampling in comparison to the old implementation?

**Reply by CarlosDerSeher** [reply]
It's actually not resampling but I take the first sample of the chunk and insert it additionally to the actual chunk or I remove the last sample of the chunk to keep up sync. Not sure yet if these artifacts are audible or not though
